services/queued/internal/queue/queue.go

```diff
@@ -228,249 +234,279 @@ func zipOutputToBase64(output_path string) (string, error) {
 	return base64.StdEncoding.EncodeToString(data), nil
 }
 
-func processMessages(wg *sync.WaitGroup) {
+// Exit causes the virtual machine to exit. Exit never returns to the caller.
+func Exit() {
+	var err error
+	err = syscall.Reboot(syscall.LINUX_REBOOT_CMD_POWER_OFF)
+	if err != nil {
+		log.Printf("power off failed: %v", err)
+	}
+	err = syscall.Reboot(syscall.LINUX_REBOOT_CMD_HALT)
+	if err != nil {
+		log.Printf("halt failed: %v", err)
+	}
+	os.Exit(1)
+	select {}
+}
+
+func ProcessMessages(wg *sync.WaitGroup, ctx context.Context) {
+
 	defer wg.Done() // let main know we are done cleaning up
 
+	var lastGoodRead = time.Now()
+
+	args := GetArguments()
+
 	for {
-		msg, more := <-queue
-		if !more {
-			log.Println("<6>Message queue empty and closed.")
+
+		select {
+
+		case <-ctx.Done(): // if cancel() execute
+			log.Println("Gracefully canceling ProcessMessages loop...")
 			return
-		}
 
-		log.Println("<6>Start processing of message.")
+		default:
 
-		func() {
-			// This is to always log last request, in case something went terribly bad
-			defer func() {
-				r := recover()
-				if r != nil {
-					log.Panicf("<3>Panicing on message %v", msg.MetaData) // just fail quickly
-				} else {
-					log.Println("<6>Finished processing of message.")
-				}
-			}()
+			log.Println("Get message via http request")
 
-			var outputVolume docker.DockerVolume
-			// we do not want to delte the tmp files, it may be useful,
-			// also the line below will not do the trick, since args
-			// are bound when defer is called, not when the function is
-			// called, so it will call Cleanup on empty path :(
-			//defer outputVolume.Cleanup()
-
-			result := func() string {
-				ctx := context.Background()
-
-				config := docker.DockerConfig{
-					Volumes:     []docker.DockerVolume{},
-					DockerImage: msg.DockerImage,
-					Timeout:     uint16(msg.MaxRunTime),
-					Memory:      int64(msg.Memory * 1024 * 1024),
-					Username:    arguments.DockerUsername,
-					Password:    arguments.DockerPassword,
-				}
+			resp, err := http.Get(args.MqReadServiceAddr)
+			if err != nil {
+				log.Println(err)
+				continue
+			}
 
-				docker.PullImage(ctx, config)
+			msg := requestMessage{
+				MaxRunTime: 500, // 300 for test and 200 for build
+				Memory:     2048,
+			}
 
-				env, err := docker.GetDockerEnv(ctx, msg.DockerImage)
-				if err != nil {
-					return err.Error()
-				}
+			var isOk = true
```

Comment:
I think this is not very go like. The anonymous function below should return msg, bool pair, what I do not like, is that you try to parse body even if error is returned.

Reply:
Probably is not. I just have seen too little Go language code yet. What I trying to achieve here is just to log error in case it pop up while reading from web service which delivers TestRuns cases. My experience with distributed/cloud native web apps is that sometime some parts/apps/pods/ just stop working for whatever reason, either for a few seconds, or completely. In such case I like to just log errors, but do not stop/break app while it itself do not have a problem. In some time either Kubernetes or some admin restarts broken app, or just rollback it to previous working version, or some obstacles (like unreliable network) go away, and everything starts work again. 
So I like to see how such thing should be implemented in a Go way. If you can show me or point me to some good example, it would be appreciated.